README.md
# userga

一个可以不用密码的用户系统。默认只需要输入email即可注册或登录，如果需要密码，可以使用Google Authenticator扫描邮件中的二维码，使用GA生成的一次性密码登录。

## 基本功能

1. 注册：输入email即完成注册或登录。提示：本网站可以不用设置密码，直接使用email登录，如果需要设置密码，请收取邮件，并按邮件中的指示操作。
1. 登录：未设置密码：输入email直接登录。已设置密码：输入email后弹出密码提示，需要输入GA生成的一次性密码登录。
1. 设置密码：邮件内容：祝贺您成功注册本网站，您现在可以直接使用email登录网站

## 安装

    cd /usr/ports/databases/py-sqlite3 # for FreeBSD
    sudo make PYTHON_VERSION=python3.4 install  # for FreeBSD
    pip install -r requirements.txt
    python userga.py

## 配置

配置文件为当前目录下的config.json，内容为：

    {
        "db_url": "sqlite:///userga.dat",
        "web_path": "userga",
        "web_addr": "127.0.0.1",
        "web_port": 8001,
        "debug": True,
    }


## 依赖

* python 3.4+（其它版本未测试）
* bottle, mako, beaker, sqlalchemy, bottle-sqlalchemy
* [bottle-plugins](https://github.com/raptorz/bottle-plugins)
* 可选webserver前端（Apache/Nginx...）

## 贡献

程序代码中包含以下第三方前端库：

* jquery
* bootstrap

## 安全性

* 安全性不高：
* secret目前用的是base64编码随机字符后取10位，密钥空间为64^10
* OTP code只有6位数字，在30秒内试完其实是可能的
* 所以必须配合重试限制，目前是简单地限制30秒内只允许尝试一次（重置密码是5分钟）


ChangeLog.txt
## 2016-02-04

* Version 0.1


mail.py
# -*- coding: UTF-8 -*
'''
    email utils
    ~~~~~~~~~~~~~~~~
    build an email contains qrcode and send it.

    :copyright: 20160211 by raptor.zh@gmail.com.
'''
from io import BytesIO
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.mime.image import MIMEImage
import smtplib
import pyqrcode

from config import config

import logging

logger = logging.getLogger(__name__)


register_text = """
您好，
    欢迎注册本网站。
    为方便起见，本网站设计为可以不必设置和输入密码，您只需要使用email即可登录。
    如果您认为需要更加安全，也可以按如下操作启用密码功能（24小时内有效，过期请登录网站，在设置页面点击重新发送密码设置邮件）：
    首先，通过Android或iOS手机或平板电脑在Google Play或App Store下载安装免费的Google Authenticator软件：
    Android版：https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2
    iOS版：https://itunes.apple.com/us/app/google-authenticator/id388497605
    然后运行这个软件，在其中选择添加－扫描条形码，然后扫描下面这个二维码：
    %(qrurl)s
    如果无法打开这个二维码链接，也可以选择添加－手动输入验证码，然后输入您的email地址和下面这个验证码：
    %(secret)s
    添加完成后点击下面的链接：
    %(reset_url)s
    按提示输入生成的6位密码并提交，即可完成密码功能启用设置，以后登录本网站时就需要输入密码，每次的密码都由此APP临时生成，仅供一次性使用。
    再次感谢您注册本网站。
"""

register_html = """
<!DOCTYPE HTML>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>密码设置说明</title>
</head>
<body>
<p>您好，</p>
<p>欢迎注册本网站。</p>
<p>为方便起见，本网站设计为可以不必设置和输入密码，您只需要使用email即可登录。</p>
<p>如果您认为需要更加安全，也可以按如下操作启用密码功能（24小时内有效，过期请登录网站，在设置页面点击重新发送密码设置邮件）：</p>
<p>首先，通过Android或iOS手机或平板电脑在Google Play或App Store下载安装免费的Google Authenticator软件：</p>
<p><a href="https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2">Android版</a></p>
<p><a href="https://itunes.apple.com/us/app/google-authenticator/id388497605">iOS版</a></p>
<p>然后运行这个软件，在其中选择添加－扫描条形码，然后扫描下面这个二维码：</p>
<p><img src="cid:image1" /></p>
<p>添加完成后点击下面的链接：</p>
<p><a href="%(reset_url)s">设置密码</a></p>
<p>按提示输入生成的6位密码并提交，即可完成密码功能启用设置，以后登录本网站时就需要输入密码，每次的密码都由此APP临时生成，仅供一次性使用。</p>
<p>再次感谢您注册本网站。</p>
</body>
</html>
"""

resetpw_text = """
您好，
    您收到此邮件是因为有人申请重新设置此账号的密码，如果不是您本人的操作，请忽略本邮件。
    如您确实需要重设密码，请按以下操作重设密码（24小时内有效，过期请重新申请密码重置）：
    首先，通过Android或iOS手机或平板电脑在Google Play或App Store下载安装免费的Google Authenticator软件：
    Android版：https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2
    iOS版：https://itunes.apple.com/us/app/google-authenticator/id388497605
    然后运行这个软件，在其中选择添加－扫描条形码，然后扫描下面这个二维码：
    %(qrurl)s
    如果无法打开这个二维码链接，也可以选择添加－手动输入验证码，然后输入您的email地址和下面这个验证码：
    %(secret)s
    添加完成后点击下面的链接：
    %(reset_url)s
    按提示输入生成的6位密码并提交，即可完成密码功能的重设，以后登录本网站时就需要输入这个APP生成的密码，以前如果在其它手机或平板电脑上设置过本账号的密码，那些将失效，以本次设置为准。
    再次感谢您使用本网站。
"""

resetpw_html = """
<!DOCTYPE HTML>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>密码重置说明</title>
</head>
<body>
<p>您好，</p>
<p>您收到此邮件是因为有人申请重新设置此账号的密码，如果不是您本人的操作，请忽略本邮件。</p>
<p>如您确实需要重设密码，请按以下操作重设密码（24小时内有效，过期请重新申请密码重置）：</p>
<p>首先，通过Android或iOS手机或平板电脑在Google Play或App Store下载安装免费的Google Authenticator软件：</p>
<p><a href="https://play.google.com/store/apps/details?id=com.google.android.apps.authenticator2">Android版</a></p>
<p><a href="https://itunes.apple.com/us/app/google-authenticator/id388497605">iOS版</a></p>
<p>然后运行这个软件，在其中选择添加－扫描条形码，然后扫描下面这个二维码：</p>
<p><img src="cid:image1" /></p>
<p>添加完成后点击下面的链接：</p>
<p><a href="%(reset_url)s">设置密码</a></p>
<p>按提示输入生成的6位密码并提交，即可完成密码功能的重设，以后登录本网站时就需要输入这个APP生成的密码，以前如果在其它手机或平板电脑上设置过本账号的密码，那些将失效，以本次设置为准。</p>
<p>再次感谢您使用本网站。</p>
</body>
</html>
"""


def generate_qrcode(uri):
    qrc = pyqrcode.create(uri, error="M")
    buf = BytesIO()
    qrc.png(buf, scale=8)
    buf.seek(0)
    return buf


def build_email(addr_from, addr_to, mail_type, secret, uri, qrurl, reset_url):
    data = {"secret": secret, "qrurl": qrurl, "reset_url": reset_url}
    if mail_type=="register":
        subject = "密码设置说明"
        content_text = register_text % data
        content_html = register_html % data
    else:
        subject = "密码重置说明"
        content_text = resetpw_text % data
        content_html = resetpw_html % data
    
    msgRoot = MIMEMultipart('related')
    msgRoot['Subject'] = subject
    msgRoot['From'] = addr_from
    msgRoot['To'] = addr_to
    msgRoot.preamble = 'This is a multi-part message in MIME format.'

    msgAlternative = MIMEMultipart('alternative')
    msgRoot.attach(msgAlternative)

    msgText = MIMEText(content_text)
    msgAlternative.attach(msgText)

    msgText = MIMEText(content_html, 'html')
    msgAlternative.attach(msgText)

    msgImage = MIMEImage(generate_qrcode(uri).read())

    msgImage.add_header('Content-ID', '<image1>')
    msgRoot.attach(msgImage)
    return msgRoot


def sendmail(addr_to, secret, uri, qrurl, reset_url, mail_type):
    msgRoot = build_email(config['smtp_user'], addr_to, mail_type, secret, uri, qrurl, reset_url)
    smtp = smtplib.SMTP()
    logger.debug(smtp.connect(config['smtp_server'], config['smtp_port']))
    logger.debug(smtp.ehlo())
    logger.debug(smtp.starttls())
    logger.debug(smtp.ehlo())
    logger.debug(smtp.login(config['smtp_user'], config['smtp_pass']))
    logger.debug(smtp.sendmail(config['smtp_user'], addr_to, msgRoot.as_string()))
    logger.debug(smtp.quit())


model.py
# -*- coding: UTF-8 -*
'''
    data model
    ~~~~~~~~~~~~~~~~
    sqlalchemy data model.

    :copyright: 20160204 by raptor.zh@gmail.com.
'''
from config import config

from sqlalchemy import create_engine

engine = create_engine(config["db_url"])

from sqlalchemy import Table, ForeignKey, Column, String, Unicode, Float
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship, backref


Base = declarative_base()

class User(Base):
    __tablename__ = "gauser"

    email     = Column(Unicode(100), primary_key=True)
    secret    = Column(String(10), nullable=False)  # if none, password is not required
    expires   = Column(Float)
    key       = Column(String(33), nullable=False, unique=True)
    resetpw   = Column(String(10))
    inv_setpw = Column(Float)
    inv_login = Column(Float)
    inv_reset = Column(Float)


metadata = Base.metadata


if __name__ == "__main__":
    metadata.create_all(engine)


saprovider.py
# -*- coding: utf-8 -*-
"""
    saprovider
    ~~~~~~~~~~~~~~~~
    sqlalchemy data provider.

    :copyright: 20160223 by raptor.zh@gmail.com
"""
import time
from os import urandom
from urllib.parse import urlencode
from base64 import b64encode
from uuid import uuid4

from config import config

from model import User

from otpauth import OtpAuth

from mail import sendmail

import logging

logger = logging.getLogger(__name__)


def get_secret():
    return b64encode(urandom(7)).decode('utf-8')[:10]


def get_user(orm, email):
    user = orm.query(User).filter(User.email==email).first()
    return user


def add_user(orm, email):
    user = User(email=email, secret=get_secret(),
            expires=time.time()+86400,
            key=str(uuid4()).replace("-", ""))
    orm.add(user)
    orm.commit()
    return user


def send_email(orm, user, reset_url, mail_type):
    """
    mail_type: register ro resetpw
    """
    secret = user.secret if mail_type=="register" else user.resetpw
    logger.debug(secret)
    auth = OtpAuth(secret)
    uri = auth.to_uri('totp', user.email, 'userga')
    qrurl = "?".join(['https://chart.googleapis.com/chart', urlencode({'cht': 'qr', 'chs': '200x200', 'chl': uri})])
    logger.debug(qrurl)
    logger.debug(mail_type)
    sendmail(user.email, secret, uri, qrurl, reset_url, mail_type)


def get_user_by_key(orm, key):
    user = orm.query(User).filter(User.key==key).first()
    if user and user.expires and user.expires>time.time():
        return user
    return None


def set_password(orm, email, code):
    user = get_user(orm, email)
    if not user or (user.inv_setpw and user.inv_setpw>time.time()):
        logger.debug("interval error")
        return False
    user.inv_setpw = time.time()+30
    secret = user.resetpw if user.resetpw else user.secret
    auth = OtpAuth(secret)
    logger.debug(auth.valid_totp(code))
    if user.expires and user.expires>time.time() and auth.valid_totp(code):
        user.inv_setpw = None
        if user.resetpw:
            user.secret = user.resetpw
            user.resetpw = None
        user.expires = None
        orm.commit()
        return True
    orm.commit()
    return False


def check_password(orm, user, code):
    if not user or (user.inv_login and user.inv_login>time.time()):
        logger.debug("interval error")
        return False
    user.inv_login = time.time()+30
    auth = OtpAuth(user.secret)
    result = False
    if auth.valid_totp(code):
        user.inv_login = None
        result = True
    orm.commit()
    return result


def reset_password(orm, email):
    user = get_user(orm, email)
    if not user or (user.inv_reset and user.inv_reset>time.time()):
        logger.debug("interval error")
        return False
    user.inv_reset = time.time() + 300
    user.expires = time.time() + 86400
    user.resetpw = get_secret()
    user.key = str(uuid4()).replace("-", "")
    orm.commit()
    return True


config.py
# -*- coding: utf-8 -*-
"""
    default config file

    :copyright: 20160204 by raptor.zh@gmail.com.
"""
#from __future__ import unicode_literals
import sys
PY3=sys.version>"3"

from os.path import dirname, abspath, expanduser, join as joinpath
import json

import logging

logger = logging.getLogger(__name__)


config_default = {
    "db_url": "sqlite:///userga.dat",
    "web_path": "userga",
    "web_addr": "127.0.0.1",
    "web_port": 8001,
    "smtp_server": "",
    "smtp_port": 25,
    "smtp_user": "",
    "smtp_pass": "",
    "debug": True,
}


def get_fullname(*args):
    root = dirname(abspath(__file__))
    return joinpath(root, joinpath(*args)) if len(args) > 0 else root


def uniencode(s, coding="utf-8"):
    return s.encode(coding) if s and (PY3 or not isinstance(s, str)) else s


def unidecode(s, coding="utf-8"):
    return unicode(s, coding) if s and (not PY3 or isinstance(s, str)) else s


try:
    with open(get_fullname("config.json"), "r") as f:
        config = json.loads(f.read())
    config_default.update(config)
    config = config_default
except IOError:
    config = config_default
